commit 5534af5fd97b8db28d4d882c62ac14efc6e43bbb
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Sat Nov 23 11:20:37 2024 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Sat Nov 23 11:20:37 2024 -0500

    Support Aux Message and Object Streaming in SP and CP (#3068)
    
    * support SP/CP aux msg
    
    * resolve merge conflict
    
    * resolve conflict
    
    * reformat
    
    * fix process type setting
    
    * add streaming to CP and SP
    
    * removed unused args
    
    * remove unsed self.fl_ctx_mgr
    
    * address PR comments
    
    * reformat
---
 nvflare/private/fed/client/client_engine.py | 205 +++++++++++++++++++++++++++-
 nvflare/private/fed/server/server_engine.py |  91 ++++++++++--
 2 files changed, 280 insertions(+), 16 deletions(-)

diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index 6f293abd..c91b9e68 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -18,17 +18,26 @@ import re
 import shutil
 import sys
 import threading
+from typing import List
 
 from nvflare.apis.event_type import EventType
 from nvflare.apis.fl_component import FLComponent
-from nvflare.apis.fl_constant import FLContextKey, MachineStatus, SystemComponents, WorkspaceConstants
+from nvflare.apis.fl_constant import FLContextKey, MachineStatus, ProcessType, SystemComponents, WorkspaceConstants
 from nvflare.apis.fl_context import FLContext, FLContextManager
+from nvflare.apis.shareable import Shareable
+from nvflare.apis.streaming import ConsumerFactory, ObjectProducer, StreamableEngine, StreamContext
+from nvflare.apis.utils.fl_context_utils import gen_new_peer_ctx
 from nvflare.apis.workspace import Workspace
-from nvflare.private.defs import ERROR_MSG_PREFIX, ClientStatusKey, EngineConstant
+from nvflare.fuel.f3.cellnet.cell import Cell
+from nvflare.fuel.f3.cellnet.defs import CellChannel, MessageHeaderKey, ReturnCode
+from nvflare.fuel.f3.message import Message as CellMessage
+from nvflare.private.aux_runner import AuxMsgTarget, AuxRunner
+from nvflare.private.defs import ERROR_MSG_PREFIX, ClientStatusKey, EngineConstant, new_cell_message
 from nvflare.private.event import fire_event
 from nvflare.private.fed.server.job_meta_validator import JobMetaValidator
 from nvflare.private.fed.utils.app_deployer import AppDeployer
 from nvflare.private.fed.utils.fed_utils import security_close
+from nvflare.private.stream_runner import ObjectStreamer
 from nvflare.security.logging import secure_format_exception, secure_log_traceback
 
 from .client_engine_internal_spec import ClientEngineInternalSpec
@@ -45,8 +54,8 @@ def _remove_custom_path():
         sys.path.remove(path)
 
 
-class ClientEngine(ClientEngineInternalSpec):
-    """ClientEngine runs in the client parent process."""
+class ClientEngine(ClientEngineInternalSpec, StreamableEngine):
+    """ClientEngine runs in the client parent process (CP)."""
 
     def __init__(self, client: FederatedClient, args, rank, workers=5):
         """To init the ClientEngine.
@@ -64,6 +73,9 @@ class ClientEngine(ClientEngineInternalSpec):
         self.rank = rank
         self.client_executor = JobExecutor(client, os.path.join(args.workspace, "startup"))
         self.admin_agent = None
+        self.aux_runner = AuxRunner(self)
+        self.object_streamer = ObjectStreamer(self.aux_runner)
+        self.cell = None
 
         self.fl_ctx_mgr = FLContextManager(
             engine=self,
@@ -76,6 +88,7 @@ class ClientEngine(ClientEngineInternalSpec):
                 SystemComponents.FED_CLIENT: client,
                 FLContextKey.SECURE_MODE: self.client.secure_train,
                 FLContextKey.WORKSPACE_ROOT: args.workspace,
+                FLContextKey.PROCESS_TYPE: ProcessType.CLIENT_PARENT,
             },
         )
 
@@ -89,6 +102,189 @@ class ClientEngine(ClientEngineInternalSpec):
     def fire_event(self, event_type: str, fl_ctx: FLContext):
         fire_event(event=event_type, handlers=self.fl_components, ctx=fl_ctx)
 
+    def get_cell(self):
+        """Get the communication cell.
+        This method must be implemented since AuxRunner calls to get cell.
+
+        Returns:
+
+        """
+        return self.cell
+
+    def initialize_comm(self, cell: Cell):
+        """This is called when communication cell has been created.
+        We will set up aux message handler here.
+
+        Args:
+            cell:
+
+        Returns:
+
+        """
+        cell.register_request_cb(
+            channel=CellChannel.AUX_COMMUNICATION,
+            topic="*",
+            cb=self._handle_aux_message,
+        )
+        self.cell = cell
+
+    def _handle_aux_message(self, request: CellMessage) -> CellMessage:
+        assert isinstance(request, CellMessage), "request must be CellMessage but got {}".format(type(request))
+        data = request.payload
+
+        topic = request.get_header(MessageHeaderKey.TOPIC)
+        with self.new_context() as fl_ctx:
+            reply = self.aux_runner.dispatch(topic=topic, request=data, fl_ctx=fl_ctx)
+            shared_fl_ctx = gen_new_peer_ctx(fl_ctx)
+            reply.set_header(key=FLContextKey.PEER_CONTEXT, value=shared_fl_ctx)
+
+            if reply is not None:
+                return_message = new_cell_message({}, reply)
+                return_message.set_header(MessageHeaderKey.RETURN_CODE, ReturnCode.OK)
+            else:
+                return_message = new_cell_message({}, None)
+            return return_message
+
+    def register_aux_message_handler(self, topic: str, message_handle_func):
+        """Register aux message handling function with specified topics.
+
+        Exception is raised when:
+            a handler is already registered for the topic;
+            bad topic - must be a non-empty string
+            bad message_handle_func - must be callable
+
+        Implementation Note:
+            This method should simply call the ServerAuxRunner's register_aux_message_handler method.
+
+        Args:
+            topic: the topic to be handled by the func
+            message_handle_func: the func to handle the message. Must follow aux_message_handle_func_signature.
+
+        """
+        self.aux_runner.register_aux_message_handler(topic, message_handle_func)
+
+    def send_aux_request(
+        self,
+        topic: str,
+        request: Shareable,
+        timeout: float,
+        fl_ctx: FLContext,
+        optional=False,
+        secure=False,
+    ) -> Shareable:
+        """Send a request to the Server via the aux channel.
+
+        Implementation: simply calls the AuxRunner's send_aux_request method.
+
+        Args:
+            topic: topic of the request.
+            request: request to be sent
+            timeout: number of secs to wait for replies. 0 means fire-and-forget.
+            fl_ctx: FL context
+            optional: whether this message is optional
+            secure: send the aux request in a secure way
+
+        Returns: a dict of replies (client name => reply Shareable)
+
+        """
+        reply = self.aux_runner.send_aux_request(
+            targets=[AuxMsgTarget.server_target()],
+            topic=topic,
+            request=request,
+            timeout=timeout,
+            fl_ctx=fl_ctx,
+            optional=optional,
+            secure=secure,
+        )
+
+        self.logger.info(f"got aux reply: {reply}")
+
+        if len(reply) > 0:
+            return next(iter(reply.values()))
+        else:
+            return Shareable()
+
+    def stream_objects(
+        self,
+        channel: str,
+        topic: str,
+        stream_ctx: StreamContext,
+        targets: List[str],
+        producer: ObjectProducer,
+        fl_ctx: FLContext,
+        optional=False,
+        secure=False,
+    ):
+        """Send a stream of Shareable objects to receivers.
+
+        Args:
+            channel: the channel for this stream
+            topic: topic of the stream
+            stream_ctx: context of the stream
+            targets: receiving sites
+            producer: the ObjectProducer that can produces the stream of Shareable objects
+            fl_ctx: the FLContext object
+            optional: whether the stream is optional
+            secure: whether to use P2P security
+
+        Returns: result from the generator's reply processing
+
+        """
+        if not self.object_streamer:
+            raise RuntimeError("object streamer has not been created")
+
+        # We are CP: can only stream to SP
+        if targets:
+            for t in targets:
+                self.logger.debug(f"ignored target: {t}")
+
+        return self.object_streamer.stream(
+            channel=channel,
+            topic=topic,
+            stream_ctx=stream_ctx,
+            targets=[AuxMsgTarget.server_target()],
+            producer=producer,
+            fl_ctx=fl_ctx,
+            secure=secure,
+            optional=optional,
+        )
+
+    def register_stream_processing(
+        self,
+        channel: str,
+        topic: str,
+        factory: ConsumerFactory,
+        stream_done_cb=None,
+        **cb_kwargs,
+    ):
+        """Register a ConsumerFactory for specified app channel and topic.
+        Once a new streaming request is received for the channel/topic, the registered factory will be used
+        to create an ObjectConsumer object to handle the new stream.
+
+        Note: the factory should generate a new ObjectConsumer every time get_consumer() is called. This is because
+        multiple streaming sessions could be going on at the same time. Each streaming session should have its
+        own ObjectConsumer.
+
+        Args:
+            channel: app channel
+            topic: app topic
+            factory: the factory to be registered
+            stream_done_cb: the callback to be called when streaming is done on receiving side
+
+        Returns: None
+
+        """
+        if not self.object_streamer:
+            raise RuntimeError("object streamer has not been created")
+
+        self.object_streamer.register_stream_processing(
+            topic=topic, channel=channel, factory=factory, stream_done_cb=stream_done_cb, **cb_kwargs
+        )
+
+    def shutdown_streamer(self):
+        if self.object_streamer:
+            self.object_streamer.shutdown()
+
     def set_agent(self, admin_agent):
         self.admin_agent = admin_agent
 
@@ -233,6 +429,7 @@ class ClientEngine(ClientEngineInternalSpec):
         thread = threading.Thread(target=shutdown_client, args=(self.client, touch_file))
         thread.start()
 
+        self.shutdown_streamer()
         return "Shutdown the client..."
 
     def restart(self) -> str:
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 295fb31c..f9b11833 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -30,25 +30,26 @@ from nvflare.apis.fl_constant import (
     AdminCommandNames,
     FLContextKey,
     MachineStatus,
-    ReturnCode,
     RunProcessKey,
     ServerCommandKey,
     ServerCommandNames,
     SnapshotKey,
     WorkspaceConstants,
 )
-from nvflare.apis.fl_context import FLContext, FLContextManager
+from nvflare.apis.fl_context import FLContext
 from nvflare.apis.fl_snapshot import RunSnapshot
 from nvflare.apis.impl.job_def_manager import JobDefManagerSpec
 from nvflare.apis.job_def import Job
 from nvflare.apis.job_launcher_spec import JobLauncherSpec
-from nvflare.apis.shareable import Shareable, make_reply
-from nvflare.apis.streaming import ConsumerFactory, ObjectProducer, StreamContext
-from nvflare.apis.utils.fl_context_utils import get_serializable_data
+from nvflare.apis.shareable import ReturnCode, Shareable, make_reply
+from nvflare.apis.streaming import ConsumerFactory, ObjectProducer, StreamableEngine, StreamContext
+from nvflare.apis.utils.fl_context_utils import gen_new_peer_ctx, get_serializable_data
 from nvflare.apis.workspace import Workspace
-from nvflare.fuel.f3.cellnet.core_cell import FQCN
+from nvflare.fuel.f3.cellnet.cell import Cell
 from nvflare.fuel.f3.cellnet.defs import MessageHeaderKey
 from nvflare.fuel.f3.cellnet.defs import ReturnCode as CellMsgReturnCode
+from nvflare.fuel.f3.cellnet.fqcn import FQCN
+from nvflare.fuel.f3.message import Message as CellMessage
 from nvflare.fuel.utils.argument_utils import parse_vars
 from nvflare.fuel.utils.zip_utils import zip_directory_to_bytes
 from nvflare.private.admin_defs import Message, MsgHeader
@@ -76,7 +77,7 @@ from .server_engine_internal_spec import EngineInfo, ServerEngineInternalSpec
 from .server_status import ServerStatus
 
 
-class ServerEngine(ServerEngineInternalSpec):
+class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
     def __init__(self, server, args, client_manager: ClientManager, snapshot_persistor, workers=3):
         """Server engine.
 
@@ -93,7 +94,7 @@ class ServerEngine(ServerEngineInternalSpec):
         self.exception_run_processes = {}
         self.run_manager = None
         self.conf = None
-        # TODO:: does this class need client manager?
+        self.cell = None
         self.client_manager = client_manager
 
         self.widgets = {
@@ -381,10 +382,61 @@ class ServerEngine(ServerEngineInternalSpec):
         return result
 
     def set_run_manager(self, run_manager: RunManager):
+        self.logger.debug("set_run_manager is called")
         self.run_manager = run_manager
+
+        # we set the run_manager's cell if we have the cell.
+        if self.cell:
+            self.run_manager.cell = self.cell
+
         for _, widget in self.widgets.items():
             self.run_manager.add_handler(widget)
 
+    def get_cell(self):
+        return self.cell
+
+    def initialize_comm(self, cell: Cell):
+        """This is called when the communication cell has been created.
+        We will set up aux message handler here.
+
+        Args:
+            cell:
+
+        Returns:
+
+        """
+        self.logger.info("initialize_comm called!")
+        self.cell = cell
+        if self.run_manager:
+            # Note that the aux_runner is created with the self.run_manager as the "engine".
+            # We must set the cell in it; otherwise it won't be able to send messages.
+            # The timing of the creation of the run_manager and the cell is not deterministic, we set the cell here
+            # only if the run_manager has been created.
+            self.run_manager.cell = cell
+
+        cell.register_request_cb(
+            channel=CellChannel.AUX_COMMUNICATION,
+            topic="*",
+            cb=self._handle_aux_message,
+        )
+
+    def _handle_aux_message(self, request: CellMessage) -> CellMessage:
+        assert isinstance(request, CellMessage), "request must be CellMessage but got {}".format(type(request))
+        data = request.payload
+
+        topic = request.get_header(MessageHeaderKey.TOPIC)
+        with self.new_context() as fl_ctx:
+            reply = self.run_manager.aux_runner.dispatch(topic=topic, request=data, fl_ctx=fl_ctx)
+            shared_fl_ctx = gen_new_peer_ctx(fl_ctx)
+            reply.set_header(key=FLContextKey.PEER_CONTEXT, value=shared_fl_ctx)
+
+            if reply is not None:
+                return_message = new_cell_message({}, reply)
+                return_message.set_header(MessageHeaderKey.RETURN_CODE, CellMsgReturnCode.OK)
+            else:
+                return_message = new_cell_message({}, None)
+            return return_message
+
     def set_job_runner(self, job_runner: JobRunner, job_manager: JobDefManagerSpec):
         self.job_runner = job_runner
         self.job_def_manager = job_manager
@@ -401,10 +453,8 @@ class ServerEngine(ServerEngineInternalSpec):
         if self.run_manager:
             return self.run_manager.new_context()
         else:
-            # return FLContext()
-            return FLContextManager(
-                engine=self, identity_name=self.server.project_name, job_id="", public_stickers={}, private_stickers={}
-            ).new_context()
+            # this call should never be made before the run_manager is created!
+            raise RuntimeError("no run_manager in Server Engine.")
 
     def add_component(self, component_id: str, component):
         self.server.runner_config.add_component(component_id, component)
@@ -563,6 +613,12 @@ class ServerEngine(ServerEngineInternalSpec):
         optional=False,
         secure=False,
     ):
+        if not self.run_manager:
+            raise RuntimeError("run_manager has not been created")
+
+        if not self.run_manager.object_streamer:
+            raise RuntimeError("object_streamer has not been created")
+
         return self.run_manager.object_streamer.stream(
             channel=channel,
             topic=topic,
@@ -582,10 +638,20 @@ class ServerEngine(ServerEngineInternalSpec):
         stream_done_cb=None,
         **cb_kwargs,
     ):
+        if not self.run_manager:
+            raise RuntimeError("run_manager has not been created")
+
+        if not self.run_manager.object_streamer:
+            raise RuntimeError("object_streamer has not been created")
+
         self.run_manager.object_streamer.register_stream_processing(
             channel=channel, topic=topic, factory=factory, stream_done_cb=stream_done_cb, **cb_kwargs
         )
 
+    def shutdown_streamer(self):
+        if self.run_manager and self.run_manager.object_streamer:
+            self.run_manager.object_streamer.shutdown()
+
     def sync_clients_from_main_process(self):
         # repeatedly ask the parent process to get participating clients until we receive the result
         # or timed out after 30 secs (already tried 30 times).
@@ -888,6 +954,7 @@ class ServerEngine(ServerEngineInternalSpec):
 
     def close(self):
         self.executor.shutdown()
+        self.shutdown_streamer()
 
 
 def server_shutdown(server, touch_file):
